### DISCUSSION TITLE
:takeout_box: Think outside the box on box variables: Latent variable ~ Stock variable? (with biflow)

### DISCUSSION THREAD

**Discussion by hyunjimoon** [Data4DM/BayesSD#94] [people; relating]
@jandraor and I were discussing "would there be any parameter without physical constraint". I think social science or economic coefficient may be e.g. coefficient of happiness w.r.t. money an example for parameter without constraint. 

[Andrew's talk on Bayes in social science](https://statmodeling.stat.columbia.edu/2022/10/15/how-should-bayesian-methods-for-the-social-sciences-differ-from-bayesian-methods-for-the-physical-sciences-my-talk-at-the-bayesian-methods-for-the-social-sciences-workshop-21-oct-2022-paris) intrigued me, so I asked him for the summary, but the comments on the blog are insightful as well.

Andrew's summary: on social science I talked about the big challenges being:  (a) in social science there is a lot of variation between people and across situations, and (b) in social science, the things we measure (such as personality, ability, attitude, value, etc.) do not have independent existences; they are defined by their measurements.
 
Bob's counter comment: Astrophysics and particle physics also have different kinds of data. Andrew’s emphasizing the high variability and measurement challenge aspect of the data in social science in the post and claiming it’s less variable in the physical sciences. I’m not sure I buy that. I was blown away by this post by John Cook on [planetary orbits are very nearly circular](https://www.johndcook.com/blog/2022/10/13/very-nearly-circular/), which explains how Kepler took Tycho’s (apparently very accurate measurements) and worried about a discrepancy of 0.037% (!) from a circular orbit (Cook’s whole sequence of posts on this topic is fascinating).

Could this be relevant to bi-flow? When do you usually use bi-flow? @tomfid Below is example from Jorgen Rander's world model (earth4all website).

![image](https://user-images.githubusercontent.com/30194633/201505140-1da99b41-22df-4bd1-810a-ba4425d18b02.png)

**Comment by hyunjimoon** [comment]
HIDDEN
-- e.g. social science, service (job satisfaction)
-- things we measure (such as personality, ability, attitude, value) do not have independent existences; they are defined by their measurements.
-- structural difficulty faced during stanify implementation (every matching happens with stock-type variables) is something to chew on.
-- Alexander shared he recovered good reference modes even with negative stock value. 
-- Jorgen's model where indexes are modeled as stock with biflow. 

If stocky flavor of measurement is agreed, perhaps, we should look for a statistical optimization model (e.g. gaussian process) to address autocorrelation embedded in time series (afterall, flow we are using is not iid, no?)

Ways to address constraints for i) parameter  ii) stock variables (non negative) need discussion, and which can be added to #7 (extending @tomfid's decision tree for estimated parameters).

**Reply by tomfid** [reply]
I don't necessarily agree with the "stocky" nature of measurements.

Certainly the observation that a flow has no instantaneous existence, so its measurement process always involves accumulation in a reporting stock, goes all the way back to early Forrester writings.

However, this doesn't mean every measurement needs to be a stock. Examples:

1. The measurement stocks in the current inventory model are basically SMOOTHs with a one-TIME STEP lag, i.e.,
measurement = INTEG( (flow-measurement)/TIME STEP )
That doesn't really do anything for you in terms of modeling a reporting process, it just introduces an arbitrarily small lag and a lot of complexity.

2. If the model is continuous/deterministic (and maybe even if it's not) the flows are often smooth enough that they can be used directly, without bothering to model the reporting process. It's not "right" in a purist sense, but often adequate.

3. Flows might represent sampling at intervals that are small with respect to TIME STEP, so they're effectively a discrete process. This happens in our CWD model for example, because almost all deer tests occur in the 9-day window of hunting season. There are some lags in reporting, but for purposes of model-data comparisons it's reasonable to generate all the samples at one point.

4. A measurement might not be a stock because it's a combination of stocks. In this case, it's an auxiliary. The extreme case of this is the standard formulation of the Kalman filter, where you have:
dX = AX + EZ
Y = BX + MZ
Where X is the vector of system states, A is the system gain matrix, E is the driving noise covariance matrix operating on white noise Z, Y is the data, B is the measurement matrix combining system states, and M is measurement error covariance. The key is that you would not want to have to add extra stocks for BX terms corresponding with Y.

I'm not sure if this is a constraint of Stan or just Stanify, but I'd rather not have it.

**Reply by tomfid** [reply]
The Kalman filter in Vensim addresses autocorrelation; if things are working right, residuals should be white noise. However, it's expensive to set up and run, and therefore not much used.

**Reply by tomfid** [reply]
Great John Cook link on orbits. Love that guy.

**Reply by hyunjimoon** [reply]
> this is a constraint of Stan or just Stanify,

I think Stan. The quantities that are penalized with distribution functions can only be integrated results e.g. [here](https://github.com/Data4DM/stanify/blob/54ed5380136a5d7d6f6119d379cd569cb001f538/stan_files/pp_S1R1/pp_S1R1_data2draws.stan#L38). 

Flow is only approachable within the function block (and adding stochasticity within function block requires c++ coding, I heard). Michael explains Stan cannot solve SDE https://discourse.mc-stan.org/t/how-to-solve-stochastic-differential-equations-in-stan/1162/3

**Reply by tomfid** [reply]
In that case, rather than using (flow-flowStock)/tau with tau=DT, perhaps it would make sense to use tau = something small, but fixed and >2*DT. That way if the RK45-auto routine is slicing DT into smaller intervals, the integration will be stable. If DT is .0625, tau=.25 would probably work well, for example.

**Reply by hyunjimoon** [reply]
> you would not want to have to add extra stocks for BX terms corresponding with Y

Could you elaborate on this a little?

> when tau > 2*DT, RK45-auto routine is slicing DT into smaller intervals, the integration will be stable

May I ask why having tau = DT prevents slicing DT into smaller intervals? In stanify, vensim's `time_step` is not translated and what we used to call `time_step` is in fact `save_pars` (= .125). In this case, how would I define DT?

I had a discussion with Hazhir on how flow (flowed stock) VS stock as target simulated (e.g. production start rate and production rate VS wip and inventory) affect estimation. 
a. RQ_{stock} >  RQ_{flow} : required stocked flow structure which depends on timestep and delta (i  but with long enough data and Tom's remedy of tau > 2*DT
b. RQ_{stock} <  RQ_{flow}: narrower credible intervals in (e.g. current negativebinom) loglikelihood's failure to capture autocorrelation. A similar observation is a tight confidence interval for Kalman filtering reported in the following slide.

I wish to stick with stock as the target simulated (and search for different algorithm #98), but when we have only flow data, I don't what to do atm.

**Comment by tomfid** [comment]
> Andrew's summary: on social science I talked about the big challenges being: (a) in social science there is a lot of variation between people and across situations, and (b) in social science, the things we measure (such as personality, ability, attitude, value, etc.) do not have independent existences; they are defined by their measurements.

I think it's not so much that these things have no independent existence, as that there's no general understanding or agreement about what that existence is.

Per my comment on "stocky" nature above, we might have a measurement Y that quantifies a combination of underlying states BX, but we don't know the weights in B, and we may not even know about all the elements of X.

Physics has an advantage, in that you generally have some principled way to decide what variables are in X, even if the weights are uncertain. As soon as you start moving towards biology, things are already pretty messy though, with lots of unknown processes in addition to a few things you can pin down from conservation laws etc.

**Reply by tomfid** [reply]
> Could this be relevant to bi-flow? When do you usually use bi-flow? @tomfid Below is example from Jorgen Rander's world model (earth4all website).

These indexes look like pretty standard smoothing processes (aka adaptive expectations or goal-gap adjustment). I use them all the time. I think the uniflow/biflow distinction introduced by STELLA long ago has caused more confusion than convenience.

**Comment by tomorrowyep** [comment]
已经收到邮箱，稍后给你回复，谢谢
